packages/plugin-node-modules/sources/NodeModulesLinker.ts

```diff
@@ -636,36 +657,150 @@ const symlinkPromise = async (srcPath: PortablePath, dstPath: PortablePath) => {
   }
 };
 
-const copyPromise = async (dstDir: PortablePath, srcDir: PortablePath, {baseFs, innerLoop}: {baseFs: FakeFS<PortablePath>, innerLoop?: boolean}) => {
+async function checksumFile(path: PortablePath, baseFs: FakeFS<PortablePath>) {
+  return await new Promise<string>((resolve, reject) => {
+    const hash = crypto.createHash(`sha1`);
+    const stream = baseFs.createReadStream(path);
+
+    stream.on(`data`, chunk => {
+      hash.update(chunk);
+    });
+
+    stream.on(`error`, error => {
+      reject(error);
+    });
+
+    stream.on(`end`, () => {
+      resolve(hash.digest(`hex`));
+    });
+  });
+}
+
+async function atomicFileWriteIfNotExist(tmpDir: PortablePath, dstPath: PortablePath, content: Buffer) {
+  const doesFileExist = await xfs.existsPromise(dstPath);
+  if (!doesFileExist) {
+    const tmpPath = ppath.join(tmpDir, toFilename(`${crypto.randomBytes(16).toString(`hex`)}.tmp`));
```

Comment:
We already have a function for temporary file creation (`mktemp`), and another for atomic writes. Why add a new one?

Reply:
`mktemp`, that we have, creates files in the temporary directory, but I need to create the files in a store directory, otherwise, I will have a huge penalty on Windows, when the temporary directory is on the other drive, from the hardlinks storage directory. Which function do you refer to, for atomic writes?

Reply:
[`lockPromise`](https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-fslib/sources/FakeFS.ts#L552) (that's what we use when writing the cache file)

Reply:
`lockPromise` is not an atomic file write though, is it?